pkg/tinydtls/contrib/sock_dtls.c

```diff
@@ -389,16 +469,35 @@ ssize_t sock_dtls_recv(sock_dtls_t *sock, sock_dtls_session_t *remote,
     assert(data);
     assert(remote);
 
-    if (sock->buf != NULL) {
-        /* there is already decrypted data available */
-        return _copy_buffer(sock, data, max_len);
-    }
-
     /* loop breaks when timeout or application data read */
     while (1) {
+        ssize_t res;
         uint32_t start_recv = xtimer_now_usec();
-        ssize_t res = sock_udp_recv(sock->udp_sock, data, max_len, timeout,
-                                    &remote->ep);
+        msg_t msg;
+
+        if (sock->buf != NULL) {
+            return _copy_buffer(sock, remote, data, max_len);
+        }
+        else if (mbox_try_get(&sock->mbox, &msg) &&
+                 msg.type == DTLS_EVENT_CONNECTED) {
+            memcpy(&remote->dtls_session, msg.content.ptr, sizeof(session_t));
+            if (sock->async_cb) {
+                sock_async_flags_t flags = SOCK_ASYNC_CONN_RDY;
+
+                if (cib_avail(&sock->mbox.cib)) {
+                    if (sock->buf) {
+                        flags |= SOCK_ASYNC_MSG_RECV;
+                    }
+                    else {
+                        flags |= SOCK_ASYNC_CONN_RECV;
+                    }
+                }
+                sock->async_cb(sock, flags, sock->async_cb_arg);
+            }
+            return -SOCK_DTLS_HANDSHAKE;
+        }
```

Comment:
So this block is waiting for the initial handshake to complete?

Reply:
Yes.

Reply:
Want to add a comment? It's always easier to navigate the code when there are some breadcrumbs every once in a while :wink:

Reply:
Following my own credo "if you can soummarize the function of a block in a comment, make the block a function", I move it to an appropriately named static function instead :-).